File: workaholic.ts
import {
  CommandContext,
  CommandHandlerResult,
  makeCommand,
} from "./command_utils.ts";
import {
  APIApplicationCommandInteractionDataSubcommandOption,
  APIMessage,
  APIUser,
  ApplicationCommandOptionType,
  markdownTable,
} from "./deps.ts";
import { getGuildMembers, getMessagesFromChannel } from "./discord_api.ts";

enum WorkaholicSubcommandOption {
  Add = "add",
  Check = "check",
}

enum WorkaholicAddCommandOption {
  What = "what",
  When = "when",
  Duration = "duration",
  Type = "type",
}

export enum WorkaholicType {
  Overtime = "OT",
  PriorityHours = "PH",
}

const workaholicAddCommandDefaultValue = {
  [WorkaholicAddCommandOption.Type]: WorkaholicType.Overtime,
};

enum WorkaholicCheckCommandOption {
  Who = "who",
  When = "when",
}

enum DebugCommandOption {
  NoMonthMatching = "no-month-matching",
}

export const workaholicCommand = makeCommand({
  name: "workaholic",
  description: "Add or check workaholic points",
  options: [
    {
      name: WorkaholicSubcommandOption.Add,
      description: "Add workaholic point for yourself",
      type: ApplicationCommandOptionType.Subcommand,
      options: [
        {
          name: WorkaholicAddCommandOption.What,
          description: "What did you do?",
          type: ApplicationCommandOptionType.String,
          required: true,
        },
        {
          name: WorkaholicAddCommandOption.When,
          description: "When did you do it?",
          type: ApplicationCommandOptionType.String,
          required: true,
        },
        {
          name: WorkaholicAddCommandOption.Duration,
          description: "How long did you do it (in hours)?",
          type: ApplicationCommandOptionType.Integer,
          required: true,
        },
        {
          name: WorkaholicAddCommandOption.Type,
          description: `What's the workaholic type? Default to ${
            workaholicAddCommandDefaultValue[WorkaholicAddCommandOption.Type]
          }`,
          type: ApplicationCommandOptionType.String,
          required: false,
          choices: [
            {
              name: WorkaholicType.Overtime,
              value: WorkaholicType.Overtime,
            },
            {
              name: WorkaholicType.PriorityHours,
              value: WorkaholicType.PriorityHours,
            },
          ],
        },
      ],
    },
    {
      name: WorkaholicSubcommandOption.Check,
      description: "Check scoreboard",
      type: ApplicationCommandOptionType.Subcommand,
      options: [
        {
          name: WorkaholicCheckCommandOption.Who,
          description: "For who? Default to everyone.",
          type: ApplicationCommandOptionType.User,
          required: false,
        },
        {
          name: WorkaholicCheckCommandOption.When,
          description: "(NOT IMPLEMENTED) For when? Default to current month.",
          type: ApplicationCommandOptionType.String,
          required: false,
        },
      ],
    },
  ],
});

type RawWorkaholicToAdd = {
  what: unknown;
  when: unknown;
  duration: unknown;
  type: unknown;
};

function sanitizeWorkaholicAdd(
  { what, when, duration, type: rType }: RawWorkaholicToAdd,
) {
  const sType = String(rType);
  return {
    what: String(what),
    when: String(when),
    duration: typeof duration === "number" ? duration : 0,
    type:
      (Object.values(WorkaholicType).includes(sType as WorkaholicType)
        ? sType
        : workaholicAddCommandDefaultValue[
          WorkaholicAddCommandOption.Type
        ]),
  };
}

function formatUserIdForDiscord(userId: string) {
  return `<@${userId}>`;
}

const prefix = "🐴";
// https://www.ascii-code.com/character/%E2%90%9F
const separator = " \u2043 ";
export function formatWorkaholicAddCommand(
  raw: RawWorkaholicToAdd & { userId: APIUser["id"] },
) {
  const {
    what,
    when,
    duration,
    type,
  } = sanitizeWorkaholicAdd(raw);
  return [
    prefix,
    // Mention the user sending the command
    formatUserIdForDiscord(raw.userId),
    type,
    when,
    `${duration}h`,
    what,
  ].join(separator);
}

export type ValidWorkaholicMessage = NonNullable<
  ReturnType<typeof parseMessageForSummary>
>;
export function parseMessageForSummary(message: APIMessage) {
  const [_prefix, userMention, type, when, hDuration, ...rWhat] = message
    .content
    .split(
      separator,
    );

  const what = rWhat.join(separator);

  if (
    userMention == null || type == null || when == null || hDuration == null ||
    what.trim() === ""
  ) {
    return null;
  }

  // userMention = <@1047114663546077236>
  const userId = userMention.slice(2, -1);
  const duration = hDuration.slice(0, -1);

  return {
    userId,
    what,
    when,
    duration,
    type,
  };
}

const rgxDateAndMonth = /[0-3]?\d \w{3}/;
const rgxOt = /ot|overtime/i;
const rgxDuration = /\d?\d\s?(jam|hours?)/i;
export function isMessagePartialMatch(messageContent: APIMessage["content"]) {
  // If message contains code block,
  // assume it's a big unrelated msg and unlikely to be
  // partial match.
  if (messageContent.includes("```")) {
    return false;
  }

  const containsOt = rgxOt.test(messageContent);
  if (rgxDuration.test(messageContent)) {
    // If msg contains duration,
    // no need to check date and month again
    // because the regex overlap,
    // date n month are looser version of duration.
    // e.g. 2 jam is a matching date n month
    // can be worked around using more specific month abbrev
    // but not worth it.
    return containsOt;
  }
  return containsOt && rgxDateAndMonth.test(messageContent);
}

export function makePartialMatchWarning({ messages, getNicknameByUserId }: {
  messages: Array<APIMessage>;
  getNicknameByUserId: ReturnType<typeof makeGetNicknameByUserIdFn>;
}) {
  if (messages.length === 0) {
    return null;
  }
  return [
    `I also found ${messages.length} message(s) that might worth some workaholic points:`,
    ...messages.map((m) =>
      `- ${
        getNicknameByUserId(m.author.id) || formatUserIdForDiscord(m.author.id)
      }, ${m.content}`
    ),
  ].join("\n");
}

enum WorkaholicResponseMessage {
  NoMessages = "No workaholic detected yet, keep working!",
}

export function makeSummaryTable(
  {
    workaholicMessages,
    getNicknameByUserId,
  }: {
    workaholicMessages: Array<
      NonNullable<ReturnType<typeof parseMessageForSummary>>
    >;
    getNicknameByUserId: ReturnType<typeof makeGetNicknameByUserIdFn>;
  },
) {
  const header = [
    "Who",
    "When",
    "How Long",
    "What",
    "Type",
  ];
  const table = markdownTable(
    [
      header,
      // TODO: add sorting/grouping
      ...workaholicMessages.map((m) => [
        getNicknameByUserId(m.userId) || formatUserIdForDiscord(m.userId),
        m.when,
        m.duration,
        m.what,
        m.type,
      ]),
    ],
  );
  return [
    "```markdown",
    table,
    "```",
  ].join("\n");
}

function makeGetNicknameByUserIdFn(
  members: Awaited<ReturnType<typeof getGuildMembers>>,
) {
  const nicknameByUserId = new Map<string, string>();

  for (const member of members) {
    if (member.user) {
      nicknameByUserId.set(member.user.id, member.nick || member.user.username);
    }
  }

  return function getNicknameByUserId(userId: string) {
    return nicknameByUserId.get(userId);
  };
}

export async function handleWorkaholicCommand(
  { interactionData, user, channelId, guildId }: CommandContext,
): Promise<CommandHandlerResult> {
  const options = interactionData.options || [];

  const addCommand = options.find((
    option,
  ): option is APIApplicationCommandInteractionDataSubcommandOption =>
    option.name === WorkaholicSubcommandOption.Add &&
    option.type === ApplicationCommandOptionType.Subcommand
  );

  if (addCommand) {
    const addOptions = addCommand.options || [];
    const whatOption = addOptions.find((option) =>
      option.name === WorkaholicAddCommandOption.What
    );
    const whenOption = addOptions.find((option) =>
      option.name === WorkaholicAddCommandOption.When
    );
    const durationOption = addOptions.find((option) =>
      option.name === WorkaholicAddCommandOption.Duration
    );
    const typeOption = addOptions.find((option) =>
      option.name === WorkaholicAddCommandOption.Type
    );
    return {
      responseText: formatWorkaholicAddCommand({
        what: whatOption?.value,
        when: whenOption?.value,
        duration: durationOption?.value,
        type: typeOption?.value,
        userId: user.id,
      }),
    };
  }

  const checkCommand = options.find((
    option,
  ): option is APIApplicationCommandInteractionDataSubcommandOption =>
    option.name === WorkaholicSubcommandOption.Check &&
    option.type === ApplicationCommandOptionType.Subcommand
  );

  if (checkCommand) {
    // It might be impossible to implement the when option efficiently
    // because the API does not provide:
    // - search API for messages
    // - before or after using timestamp instead of snowflake id
    // The bruteforce way to implement this would
    // be to iterate through the messages until we found message
    // with suitable timestamp, which might take multiple API calls
    // depending on the when options and channel activity.
    const checkOptions = checkCommand.options || [];
    const whenOption = checkOptions.find((option) =>
      option.name === WorkaholicCheckCommandOption.When
    );

    const doMonthMatching = !String(whenOption?.value ?? "").includes(
      DebugCommandOption.NoMonthMatching,
    );

    const whoOption = checkOptions.find((option) =>
      option.name === WorkaholicCheckCommandOption.Who
    );

    const who = whoOption?.type === ApplicationCommandOptionType.User
      ? whoOption.value
      : null;

    const [allMessages, guildMembers] = await Promise.all([
      // TODO: Retrieve all messages in the specified month
      // Right now this retrieves the last 100 only
      getMessagesFromChannel({ channelId }),
      // Hardcoded to 50 members for now
      getGuildMembers({ guildId, limit: 50 }),
    ]);

    // TODO: Consider making this aware of the user executor timezone
    // Right now this uses GMT+0 and might show messages from different months
    // at the start and the end of month depending on TZ.
    const getYearMonthPart = (isoTimestamp: string) => isoTimestamp.slice(0, 7);
    const thisMonthPart = getYearMonthPart(new Date().toISOString());

    const thisMonthMessages = allMessages.filter((m) =>
      doMonthMatching
        ? getYearMonthPart(m.timestamp) === thisMonthPart
        // Don't filter out messages by month
        // when month matching is disabled.
        : true
    );

    if (thisMonthMessages.length === 0) {
      return {
        responseText: WorkaholicResponseMessage.NoMessages,
      };
    }

    const matchingUserMessages = thisMonthMessages.filter((m) =>
      who == null ? true : m.mentions.find((user) => user.id === who) != null
    );

    // Only enable partial matching when checking for everyone.
    const enablePartialMatching = who == null;

    const partialMatchMessages: Array<APIMessage> = [];
    const matchingMessages: Array<ValidWorkaholicMessage> = [];

    // With the assumption messages from discord are sorted with timestamp desc:
    // Go through every messages in reverse order
    // so the top most message is the oldest message
    for (const channelMessage of matchingUserMessages.toReversed()) {
      if (channelMessage.content.startsWith(`${prefix}${separator}`)) {
        const validMessage = parseMessageForSummary(channelMessage);
        if (validMessage) {
          matchingMessages.push(validMessage);
        } else if (enablePartialMatching) {
          partialMatchMessages.push(channelMessage);
        }
      } else {
        if (
          enablePartialMatching &&
          channelMessage.webhook_id == null && !channelMessage.author.bot &&
          isMessagePartialMatch(channelMessage.content)
        ) {
          partialMatchMessages.push(channelMessage);
        }
      }
    }

    const getNicknameByUserId = makeGetNicknameByUserIdFn(guildMembers);

    const summaryTable = matchingMessages.length > 0
      ? makeSummaryTable(
        {
          workaholicMessages: matchingMessages,
          getNicknameByUserId,
        },
      )
      : null;
    const partialMatchWarning = makePartialMatchWarning({
      messages: enablePartialMatching ? partialMatchMessages : [],
      getNicknameByUserId,
    });

    if (summaryTable == null) {
      return {
        responseText: partialMatchWarning != null
          ? `No exact matches found but...`.concat("\n", partialMatchWarning)
          : WorkaholicResponseMessage.NoMessages,
      };
    }

    const responseText = [summaryTable, partialMatchWarning].filter(
      Boolean,
    )
      .join("\n");

    return { responseText };
  }

  return {
    responseText: `💣 Unhandled subcommand!
raw data=${JSON.stringify(interactionData, null, 2)}`,
  };
}

File: v2/commands/workaholic.go
package commands

import (
	"fmt"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"bernard/discord"
)

const (
	workaholicPrefix    = "🐴"
	workaholicSeparator = " \u2043 " // " ⁃ " https://www.ascii-code.com/character/%E2%90%9F

	workaholicTypeOT = "OT"
	workaholicTypePH = "PH"

	noWorkaholicMessage = "No workaholic detected yet, keep working!"
)

var (
	rgxDateAndMonth = regexp.MustCompile(`[0-3]?\d \w{3}`)
	rgxOT           = regexp.MustCompile(`(?i)(?:ot|overtime)`)
	rgxDuration     = regexp.MustCompile(`(?i)\d?\d\s?(?:jam|hours?)`)
)

// workaholicAPIClient holds injectable API functions for testability.
type workaholicAPIClient struct {
	getMessages func(channelID string, limit int) ([]discord.Message, error)
	getMembers  func(guildID string, limit int) ([]discord.Member, error)
}

// handleWorkaholic is the production handler, using the real Discord API.
var handleWorkaholic = makeWorkaholicHandler(workaholicAPIClient{
	getMessages: discord.GetMessagesFromChannel,
	getMembers:  discord.GetGuildMembers,
})

// makeWorkaholicHandler returns a Handler using the provided API client.
// Tests pass a mock client; production uses the default above.
func makeWorkaholicHandler(api workaholicAPIClient) Handler {
	return func(ctx CommandContext) (CommandResult, error) {
		for _, opt := range ctx.InteractionData.Options {
			if opt.Type == discord.OptionTypeSubCommand {
				switch opt.Name {
				case "add":
					return handleWorkaholicAdd(ctx, opt.Options)
				case "check":
					return handleWorkaholicCheck(ctx, opt.Options, api)
				}
			}
		}
		return CommandResult{
			ResponseText: fmt.Sprintf("💣 Unhandled subcommand!\nraw data=%v", ctx.InteractionData),
		}, nil
	}
}

// --- add subcommand ---

func handleWorkaholicAdd(ctx CommandContext, opts []discord.InteractionDataOption) (CommandResult, error) {
	var what, when string
	duration := 0
	wType := workaholicTypeOT // default

	for _, opt := range opts {
		switch opt.Name {
		case "what":
			what = opt.StringValue()
		case "when":
			when = opt.StringValue()
		case "duration":
			duration = opt.IntValue()
		case "type":
			v := opt.StringValue()
			if v == workaholicTypeOT || v == workaholicTypePH {
				wType = v
			}
		}
	}

	return CommandResult{ResponseText: formatWorkaholicAdd(ctx.User.ID, what, when, duration, wType)}, nil
}

// formatWorkaholicAdd formats a workaholic entry message.
// Matches formatWorkaholicAddCommand in workaholic.ts.
func formatWorkaholicAdd(userID, what, when string, duration int, wType string) string {
	parts := []string{
		workaholicPrefix,
		"<@" + userID + ">",
		wType,
		when,
		fmt.Sprintf("%dh", duration),
		what,
	}
	return strings.Join(parts, workaholicSeparator)
}

// --- check subcommand ---

type workaholicEntry struct {
	userID   string
	what     string
	when     string
	duration string
	wType    string
}

// parseMessage extracts a workaholic entry from a message content string.
// Returns nil if the content is not a valid workaholic entry.
// Matches parseMessageForSummary in workaholic.ts.
func parseMessage(content string) *workaholicEntry {
	parts := strings.Split(content, workaholicSeparator)
	// Expected: [prefix, userMention, type, when, hDuration, ...what]
	if len(parts) < 6 {
		return nil
	}

	userMention := parts[1]
	wType := parts[2]
	when := parts[3]
	hDuration := parts[4]
	what := strings.Join(parts[5:], workaholicSeparator)

	if strings.TrimSpace(what) == "" {
		return nil
	}

	// userMention = "<@snowflakeId>"
	if len(userMention) < 4 {
		return nil
	}
	userID := userMention[2 : len(userMention)-1]

	// hDuration = "7h" → duration = "7"
	if len(hDuration) < 1 {
		return nil
	}
	duration := hDuration[:len(hDuration)-1]

	return &workaholicEntry{
		userID:   userID,
		what:     what,
		when:     when,
		duration: duration,
		wType:    wType,
	}
}

// isMessagePartialMatch returns true if the message looks like an unformatted workaholic entry.
// Matches isMessagePartialMatch in workaholic.ts.
func isMessagePartialMatch(content string) bool {
	if strings.Contains(content, "```") {
		return false
	}
	containsOT := rgxOT.MatchString(content)
	if rgxDuration.MatchString(content) {
		return containsOT
	}
	return containsOT && rgxDateAndMonth.MatchString(content)
}

func handleWorkaholicCheck(ctx CommandContext, opts []discord.InteractionDataOption, api workaholicAPIClient) (CommandResult, error) {
	doMonthMatching := true
	var who string

	for _, opt := range opts {
		switch opt.Name {
		case "when":
			if strings.Contains(opt.StringValue(), "no-month-matching") {
				doMonthMatching = false
			}
		case "who":
			if opt.Type == discord.OptionTypeUser {
				who = opt.StringValue()
			}
		}
	}

	// Fetch messages and guild members concurrently — matches Promise.all in workaholic.ts.
	type msgResult struct {
		messages []discord.Message
		err      error
	}
	type memResult struct {
		members []discord.Member
		err     error
	}

	msgCh := make(chan msgResult, 1)
	memCh := make(chan memResult, 1)

	go func() {
		msgs, err := api.getMessages(ctx.ChannelID, 100)
		msgCh <- msgResult{msgs, err}
	}()
	go func() {
		mems, err := api.getMembers(ctx.GuildID, 50)
		memCh <- memResult{mems, err}
	}()

	mr := <-msgCh
	memr := <-memCh
	if mr.err != nil {
		return CommandResult{}, mr.err
	}
	if memr.err != nil {
		return CommandResult{}, memr.err
	}

	// Filter by current month (UTC) — matches GMT+0 behavior in workaholic.ts.
	thisMonth := time.Now().UTC().Format("2006-01")
	var thisMonthMessages []discord.Message
	for _, m := range mr.messages {
		if !doMonthMatching || (len(m.Timestamp) >= 7 && m.Timestamp[:7] == thisMonth) {
			thisMonthMessages = append(thisMonthMessages, m)
		}
	}

	if len(thisMonthMessages) == 0 {
		return CommandResult{ResponseText: noWorkaholicMessage}, nil
	}

	// Filter by mentioned user if specified.
	var matchingUserMessages []discord.Message
	for _, m := range thisMonthMessages {
		if who == "" {
			matchingUserMessages = append(matchingUserMessages, m)
		} else {
			for _, u := range m.Mentions {
				if u.ID == who {
					matchingUserMessages = append(matchingUserMessages, m)
					break
				}
			}
		}
	}

	enablePartialMatching := who == ""
	prefixWithSep := workaholicPrefix + workaholicSeparator

	var partialMatchMessages []discord.Message
	var matchingEntries []workaholicEntry

	// Iterate in reverse (oldest first) — Discord returns messages newest first.
	for i := len(matchingUserMessages) - 1; i >= 0; i-- {
		msg := matchingUserMessages[i]
		if strings.HasPrefix(msg.Content, prefixWithSep) {
			if entry := parseMessage(msg.Content); entry != nil {
				matchingEntries = append(matchingEntries, *entry)
			} else if enablePartialMatching {
				partialMatchMessages = append(partialMatchMessages, msg)
			}
		} else if enablePartialMatching && msg.WebhookID == "" && !msg.Author.Bot && isMessagePartialMatch(msg.Content) {
			partialMatchMessages = append(partialMatchMessages, msg)
		}
	}

	nicknames := buildNicknameMap(memr.members)

	summaryTable := ""
	if len(matchingEntries) > 0 {
		summaryTable = makeSummaryTable(matchingEntries, nicknames)
	}

	partialWarning := ""
	if enablePartialMatching {
		partialWarning = makePartialMatchWarning(partialMatchMessages, nicknames)
	}

	if summaryTable == "" {
		if partialWarning != "" {
			return CommandResult{ResponseText: "No exact matches found but...\n" + partialWarning}, nil
		}
		return CommandResult{ResponseText: noWorkaholicMessage}, nil
	}

	parts := []string{summaryTable}
	if partialWarning != "" {
		parts = append(parts, partialWarning)
	}
	return CommandResult{ResponseText: strings.Join(parts, "\n")}, nil
}

// --- helpers ---

func buildNicknameMap(members []discord.Member) map[string]string {
	m := make(map[string]string, len(members))
	for _, member := range members {
		if member.User != nil {
			nick := member.Nick
			if nick == "" {
				nick = member.User.Username
			}
			m[member.User.ID] = nick
		}
	}
	return m
}

// makeSummaryTable renders workaholic entries as a fenced markdown table.
// Matches makeSummaryTable in workaholic.ts.
func makeSummaryTable(entries []workaholicEntry, nicknames map[string]string) string {
	header := []string{"Who", "When", "How Long", "What", "Type"}
	rows := [][]string{header}

	for _, e := range entries {
		who := nicknames[e.userID]
		if who == "" {
			who = "<@" + e.userID + ">"
		}
		rows = append(rows, []string{who, e.when, e.duration, e.what, e.wType})
	}

	return "```markdown\n" + markdownTable(rows) + "\n```"
}

// makePartialMatchWarning builds a warning message listing messages that look like
// unformatted workaholic entries. Returns empty string when there are no matches.
// Matches makePartialMatchWarning in workaholic.ts.
func makePartialMatchWarning(messages []discord.Message, nicknames map[string]string) string {
	if len(messages) == 0 {
		return ""
	}
	lines := make([]string, 0, len(messages)+1)
	lines = append(lines, fmt.Sprintf("I also found %d message(s) that might worth some workaholic points:", len(messages)))
	for _, m := range messages {
		who := nicknames[m.Author.ID]
		if who == "" {
			who = "<@" + m.Author.ID + ">"
		}
		lines = append(lines, fmt.Sprintf("- %s, %s", who, m.Content))
	}
	return strings.Join(lines, "\n")
}

// markdownTable renders rows as a GFM table with left-aligned, padded columns.
// Matches the output of the markdown-table npm package used in workaholic.ts.
func markdownTable(rows [][]string) string {
	if len(rows) == 0 {
		return ""
	}

	cols := len(rows[0])
	widths := make([]int, cols)
	for _, row := range rows {
		for j, cell := range row {
			if n := utf8.RuneCountInString(cell); n > widths[j] {
				widths[j] = n
			}
		}
	}

	lines := make([]string, 0, len(rows)+1)
	for i, row := range rows {
		var sb strings.Builder
		sb.WriteString("|")
		for j, cell := range row {
			sb.WriteString(" ")
			sb.WriteString(cell)
			sb.WriteString(strings.Repeat(" ", widths[j]-utf8.RuneCountInString(cell)))
			sb.WriteString(" |")
		}
		lines = append(lines, sb.String())

		if i == 0 {
			var sb strings.Builder
			sb.WriteString("|")
			for _, w := range widths {
				sb.WriteString(" ")
				sb.WriteString(strings.Repeat("-", w))
				sb.WriteString(" |")
			}
			lines = append(lines, sb.String())
		}
	}

	return strings.Join(lines, "\n")
}
